File: src/main/java/de/omilke/bankingfx/report/categories/model/CategoryOverTimeElement.java
package de.omilke.bankingfx.report.categories.model;

import javafx.beans.property.*;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.List;
import java.util.Objects;

/**
 * Created by Olli on 23.05.2017.
 */
public class CategoryOverTimeElement {

    private final StringProperty categoryName;

    private final ListProperty<BigDecimal> monthValues;
    private final boolean aggregateRow;

    private final ObjectProperty<BigDecimal> average;
    private final ObjectProperty<BigDecimal> sum;

    public CategoryOverTimeElement(String categoryName, ObservableList<BigDecimal> monthValues) {

        this(categoryName, monthValues, false);
    }

    private CategoryOverTimeElement(String categoryName, ObservableList<BigDecimal> monthValues, boolean aggregateRow) {

        this.categoryName = new SimpleStringProperty(this, null, categoryName);
        this.monthValues = new SimpleListProperty<>(this, "values", monthValues);
        this.aggregateRow = aggregateRow;

        BigDecimal summedAmount = getSum(monthValues);
        sum = new SimpleObjectProperty<>(this, null, summedAmount);

        //average based on the number of months spanned. That number includes months without expenditures but
        // excludes months that have no expenditures at all (those month would be skipped as a column)
        average = new SimpleObjectProperty<>(this, null, getAverage(summedAmount, monthValues.size()));
    }

    private static BigDecimal getSum(ObservableList<BigDecimal> values) {

        return values.stream()
                .filter(Objects::nonNull)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    private static BigDecimal getAverage(BigDecimal sum, int size) {

        return sum.divide(new BigDecimal(size), 2, RoundingMode.HALF_UP);
    }

    public String getCategoryName() {

        return categoryName.get();
    }

    public StringProperty categoryNameProperty() {

        return categoryName;
    }

    public ObservableList<BigDecimal> getMonthValues() {

        return monthValues.get();
    }

    public BigDecimal getAverage() {

        return average.get();
    }

    public ObjectProperty<BigDecimal> averageProperty() {

        return average;
    }

    public BigDecimal getSum() {

        return sum.get();
    }

    public ObjectProperty<BigDecimal> sumProperty() {

        return sum;
    }

    public boolean isAggregateRow() {

        return aggregateRow;
    }

    public static CategoryOverTimeElement buildSumOverTimeElement(List<CategoryOverTimeElement> categorySums) {

        ObservableList<BigDecimal> monthSumsOverTime = FXCollections.observableArrayList();

        //build sum by month (i. e. per element of value) over all model elements
        for (CategoryOverTimeElement categoryOverTime : categorySums) {

            ObservableList<BigDecimal> monthValues = categoryOverTime.getMonthValues();
            for (int monthIndex = 0; monthIndex < monthValues.size(); monthIndex++) {

                //if a month-column is missing, add it
                if (monthSumsOverTime.size() < monthIndex + 1) {
                    monthSumsOverTime.add(BigDecimal.ZERO);
                }

                BigDecimal presentlySummedValue = monthSumsOverTime.get(monthIndex);
                BigDecimal categoryMonthValue = monthValues.get(monthIndex);

                if (categoryMonthValue != null) {

                    //replace old sum with the current value added on top if there is a sum for this category and month
                    monthSumsOverTime.set(monthIndex, presentlySummedValue.add(categoryMonthValue));
                }
            }
        }

        return new CategoryOverTimeElement("Sum", monthSumsOverTime, true);
    }

}

File: src/main/java/de/omilke/bankingfx/report/categories/model/CategoryOverTimeElement.kt
package de.omilke.bankingfx.report.categories.model

import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * Created by Olli on 23.05.2017.
 */
class CategoryOverTimeElement {

    private val categoryName: StringProperty
    private val monthValues: ObjectProperty<List<BigDecimal?>>
    private val average: ObjectProperty<BigDecimal>
    private val sum: ObjectProperty<BigDecimal>

    val isAggregateRow: Boolean

    constructor(categoryName: String, monthValues: List<BigDecimal?>) :
            this(categoryName, monthValues, false)

    private constructor(categoryName: String, monthValues: List<BigDecimal?>, aggregateRow: Boolean) {

        this.categoryName = SimpleStringProperty(this, null, categoryName)
        this.monthValues = SimpleObjectProperty(this, "values", monthValues)

        val monthSum = getSum(monthValues)
        this.sum = SimpleObjectProperty(this, null, monthSum)
        this.average = SimpleObjectProperty(this, null, getAverage(monthSum, monthValues.size))

        isAggregateRow = aggregateRow
    }

    fun getCategoryName(): String {
        return categoryName.get()
    }

    fun categoryNameProperty(): StringProperty {
        return categoryName
    }

    fun getMonthValues(): List<BigDecimal?>? {
        return monthValues.get()
    }

    fun getAverage(): BigDecimal {
        return average.get()
    }

    fun averageProperty(): ObjectProperty<BigDecimal> {
        return average
    }

    fun getSum(): BigDecimal {
        return sum.get()
    }

    fun sumProperty(): ObjectProperty<BigDecimal> {
        return sum
    }

    companion object {

        private fun getSum(values: List<BigDecimal?>): BigDecimal {
            return values
                    .filterNotNull()
                    .fold(BigDecimal.ZERO, BigDecimal::add)
        }

        private fun getAverage(sum: BigDecimal, size: Int): BigDecimal {
            return sum.divide(BigDecimal(size), 2, RoundingMode.HALF_UP)
        }

        fun buildSumOverTimeElement(categorySums: List<CategoryOverTimeElement>): CategoryOverTimeElement {

            val monthSumsOverTime = ArrayList<BigDecimal>()

            //build sum by month (i. e. per element of value) over all model elements
            for (categoryOverTime in categorySums) {

                categoryOverTime.getMonthValues()?.let {
                    for (monthIndex in it.indices) {

                        //if a month-column is missing, add it
                        if (monthSumsOverTime.size < monthIndex + 1) {
                            monthSumsOverTime.add(BigDecimal.ZERO)
                        }

                        val presentlySummedValue = monthSumsOverTime[monthIndex]
                        val categoryMonthValue = it[monthIndex]

                        if (categoryMonthValue != null) {

                            //replace old sum with the current value added on top if there is a sum for this category and month
                            monthSumsOverTime[monthIndex] = presentlySummedValue.add(categoryMonthValue)
                        }
                    }
                }
            }

            return CategoryOverTimeElement("Sum", monthSumsOverTime, true)
        }
    }

}